src/fp/list.nim

```diff
@@ -114,6 +114,24 @@ proc foldRightF*[T, U](xs: List[T], z: () -> U, f: (T, () -> U) -> U): U =
   if xs.isEmpty: z()
   else: f(xs.head, () => xs.tail.foldRightF(z, f))
 
+proc unfoldLeft*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the left from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
+        (b, a) = f(a).get()
+        result = b ^^ result
+
+proc unfoldRight*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the right from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
+        (b, a) = f(a).get()
+        result = result.append(b.asList)
```

Comment:
`mratsim`, this Lisp example is just about mutating the last cons cell of the accumulator list. This is possible here, too (since Nim doesn't allow us to make a field immutable), but I think it would be better to keep using the immutable list abstraction for now.